### ISSUE TITLE
Make ValidationError Hashable

### ISSUE THREAD

**Issue by adksujan** [Julian/jsonschema#477]
cannot log ValidationError as an exception. Making the exceptions hashable would be great to use in standard logging libraries. Any plans to make ValidationError exception hashable before the v3.0 release??

**Comment by Julian** [comment]
Hey yeah this is one of the remaining things I'm thinking about before the release.

The issue with making them hashable is that doing so requires an API change, because they've got mutable attributes on them at the minute, so those would need to become immutable ones.

**Comment by stuarteberg** [comment]
Of course, `ValidationError` and `SchemaError` would be trivially hashable if they didn't define `__eq__`.  Is it necessary for those classes to be comparable?  I tried grepping the `jsonschema` code base for `==` and `!=`, and as far as I can tell, exceptions are never compared.  (But I could definitely be missing something.)

If comparability is not necessary, I vote for simply removing `_Error.__eq__()` and `_Error.__ne__()`.

---

>Making the exceptions hashable would be great to use in standard logging libraries.

Here's another use-case: `pytest` can't handle unhashable exceptions, so any users whose tests might raise a `ValidationError` will currently see an `INTERNALERROR` traceback from `pytest`, crashing the whole test suite.

For example:

```python
# test.py
from jsonschema import validate
def test():
    validate([], {"type": "object"})
```

```
$ pytest test.py

...

INTERNALERROR>   File "/prefix/lib/python3.6/site-packages/_pytest/_code/code.py", line 398, in exconly
INTERNALERROR>     lines = format_exception_only(self.type, self.value)
INTERNALERROR>   File "/prefix/lib/python3.6/traceback.py", line 136, in format_exception_only
INTERNALERROR>     return list(TracebackException(etype, value, None).format_exception_only())
INTERNALERROR>   File "/prefix/lib/python3.6/traceback.py", line 462, in __init__
INTERNALERROR>     _seen.add(exc_value)
INTERNALERROR> TypeError: unhashable type: 'ValidationError'
```

**Comment by Julian** [comment]
It's used in tests, yes, via `assertEqual`.

> Here's another use-case: pytest can't handle unhashable exceptions, so any users whose tests might raise a ValidationError will currently see an INTERNALERROR traceback from pytest, crashing the whole test suite.

Sounds like a `py.test` bug worth filing regardless, though presumably it's just the same underlying one (that py~34-36 broke this, and then fixed it again in 3.7). pytest seems like the sort of thing though that should be trying to fix it even on those though.

**Comment by stuarteberg** [comment]
>It's used in tests, yes, via assertEqual.

Ah, thanks for pointing that out.  Still, it's not used in many places, so I've proposed #502 to eliminate the need for comparable exceptions.

>Sounds like a `py.test` bug worth filing regardless

OK, I've submitted pytest-dev/pytest#4514, to see if they're interested in supporting unhashable exceptions.

**Comment by stuarteberg** [comment]
>presumably it's just the same underlying one (that py~34-36 broke this, and then fixed it again in 3.7)

FWIW, you're right: This is not an issue when using Python 3.7.  I'm using Python 3.6, where it is an issue.
